[SPARK-49556][SQL] Add SQL pipe syntax for the SELECT operator

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for the SELECT operator.

For example:

```
CREATE TABLE t(x INT, y STRING) USING CSV;
INSERT INTO t VALUES (0, 'abc'), (1, 'def');

TABLE t
|> SELECT x, y

0	abc
1	def

TABLE t
|> SELECT x, y
|> SELECT x + LENGTH(y) AS z

3
4

(SELECT * FROM t UNION ALL SELECT * FROM t)
|> SELECT x + LENGTH(y) AS result

3
3
4
4

TABLE t
|> SELECT sum(x) AS result

Error: aggregate functions are not allowed in the pipe operator |> SELECT clause; please use the |> AGGREGATE clause instead
```

### Why are the changes needed?

The SQL pipe operator syntax will let users compose queries in a more flexible fashion.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds a few unit test cases, but mostly relies on golden file test coverage. I did this to make sure the answers are correct as this feature is implemented and also so we can look at the analyzer output plans to ensure they look right as well.

### Was this patch authored or co-authored using generative AI tooling?

No

dtenedor committed:
commit

commit

commit

dtenedor committed:
add more testing

dtenedor committed:
add more testing

dtenedor commented:
cc @cloud-fan @srielau @gengliangwang

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -4989,6 +4989,15 @@ object SQLConf {
+        "the sequence of steps that the query performs in a composable fashion.")
+      .version("4.0.0")
+      .booleanConf
+      .createWithDefault(false)
```
Will the default value become true once the feature is completed?

dtenedor replied:
Yes. I plan to implement the different SQL clauses and then enable to true by default. It will be fully backwards-compatible with existing queries.

gengliangwang commented on common/utils/src/main/resources/error/error-conditions.json:
```diff
@@ -3707,6 +3707,12 @@
   },
+  "PIPE_OPERATOR_SELECT_CONTAINS_AGGREGATE_FUNCTION" : {
+    "message" : [
+      "Aggregate function <expr> is not allowed when using the pipe operator |> SELECT clause; please use the pipe operator |> AGGREGATE clause instead"
```
This is a bit odd since `|> AGGREGATE` is not added yet

dtenedor replied:
Yes, this will be tracked in a separate Jira task. The intention is that `|> SELECT` will not allow aggregate functions (ever). So this restriction on not allowing aggregate functions in `|> SELECT` will be permanent.

Until we implement `|> AGGREGATE` later, it won't be possible to use pipe operators to perform aggregation. But the feature is not complete (or enabled by default) until all the operators are implemented.

gengliangwang commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -0,0 +1,84 @@
+
+-- Selecting attributes.
+table t
+|> select x, y;
```
nit: let's add a test case with LCA?

dtenedor replied:
There is one on L64 :)

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -5677,4 +5677,29 @@ class AstBuilder extends DataTypeAstBuilder
+        havingClause = null,
+        windowClause = null,
+        left) match {
+        case p: Project =>
```
Shall we check the condition when the input is not a Project?

dtenedor replied:
Good Q, this input should always be a projection since we only pass a context for the SELECT clause here and pass "null" for all other clauses. I added a comment here to this effect.

gengliangwang replied:
Since `withSelectQuerySpecification` returns a `LogicalPlan`, we should handle the case when the result is not `Project`

dtenedor replied:
Sounds good, I put a pattern match for this with an internal error for now.

dtenedor committed:
exclude flaky ThriftServerQueryTestSuite for new golden file

exclude flaky ThriftServerQueryTestSuite for new golden file

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @gengliangwang for your review!

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2727,6 +2727,12 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
       t => t.containsAnyPattern(AGGREGATE_EXPRESSION, PYTHON_UDF) && t.containsPattern(PROJECT),
       ruleId) {
       case Project(projectList, child) if containsAggregates(projectList) =>
+        if (child.isInstanceOf[PipeOperatorSelect]) {
```
shall we move this to `CheckAnalysis`?

dtenedor replied:
Good Q, by that point `RemovePipeOperators` will have removed the `PipeOperatorSelect` so we won't be able to tell anymore, so we have to check it here.

gengliangwang replied:
We can move `RemovePipeOperators` to `FinishAnalysis`

dtenedor replied:
I tried this but it was more complex than I had hoped. I ended up switching the implementation to make the `PipeOperatorSelect` a temporary expression instead, that checks its children for aggregate functions at the end of analysis. This seems to work without creating any new analyzer rule.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @gengliangwang for your review again!

dtenedor committed:
sync

dtenedor committed:
switch to expression

switch to expression

switch to expression

switch to expression

moving error checking to checkanalysis

dtenedor reviewed (commented):
Thanks again @gengliangwang for your careful attention

cloud-fan commented on sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:
```diff
@@ -604,6 +604,7 @@ queryTerm
         operator=INTERSECT setQuantifier? right=queryTerm                                #setOperation
     | left=queryTerm {!legacy_setops_precedence_enabled}?
         operator=(UNION | EXCEPT | SETMINUS) setQuantifier? right=queryTerm              #setOperation
+    | left=queryTerm OPERATOR_PIPE operatorPipeRightSide                                 #operatorPipeStatement
```
I had a hard time understanding this recursive parser rule. How does it match continuous pipe operators? And what is the Operator Precedence with mixed classic SQL query syntax and the new pipe syntax?

cloud-fan replied:
I'm not familiar with ANTLR enough. So this recursive parser rule matches the SQL string from the end? e.g. it finds the first `operatorPipeRightSide` from the end, and then tries to match a chain of pipe operators.

dtenedor replied:
Sure, no problem, I can try to explain it.

ANTLR tokenizes each SQL query it receives, converting the input string into a sequence of tokens (using `SqlBaseLexer.g4`). Then the parser's job (in this file) is to convert that sequence of tokens into an initial unresolved logical plan representing the parse tree.

To do so, the parser checks each rule in the listed sequence, one-by-one, comparing the provided tokens at the current index in the sequence with the required tokens from the rule. If the rule matches, wherein all keywords and other components in the rule map to corresponding input tokens, then the parser generates the rule's unresolved logical plan tree using the logic in `AstBuilder.scala`.

In this case, we define the new token `OPERATOR_PIPE: '|>';` in `SqlBaseLexer.g4`. Then we add a new option to the existing `queryTerm` rule to allow any syntax matching an existing `queryTerm` to appear on the left side of this `|>` token and the syntax of `operatorPipeRightSide` on the right side (which in this PR is limited to only a `selectClause`).

ANTLR grammar allows left-recursive rules wherein any alternative may begin with a reference to the same rule, so the `queryTerm` on the left side may match any valid existing syntax for a `queryTerm` such as `TABLE t`, a table subquery, etc. Since we are extending `queryTerm` to also match against `queryTerm OPERATOR_PIPE operatorPipeRightSide`, this alternative implements the recursion wherein we may chain multiple pipe operators together. For example, in `TABLE t |> SELECT x |> LIMIT 2`, `TABLE t` matches a `queryTerm`, then `TABLE t |> SELECT x` matches another, and finally the entire query (using the new recursive `#operatorPipeStatement` alternative  two times).

Otherwise, if the rule does not match, then the parser moves on to try the next rule in the sequence, and so on, similar to a Scala pattern-match. This defines the precedence of the rules amongst each other: the ones appearing first in the list in `SqlBaseParser.g4` apply first.

xinrong-meng replied:
So the parser generates a basic parse tree, and `AstBuilder.scala` transforms that into an unresolved logical plan? Thanks for the clear and detailed explanation! I'm adding SQL syntax too and this is very helpful.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PipeSelect.scala:
```diff
@@ -0,0 +1,47 @@
+  extends UnaryExpression with RuntimeReplaceable {
+  final override val nodePatterns: Seq[TreePattern] = Seq(PIPE_OPERATOR_SELECT, RUNTIME_REPLACEABLE)
+  override def withNewChildInternal(newChild: Expression): Expression = PipeSelect(newChild)
+  override def replacement: Expression = {
```
should it be a lazy val? `replacement` may be accessed many times during analysis (get data type, nullability, etc.)

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -5711,6 +5711,47 @@ class AstBuilder extends DataTypeAstBuilder
+      def updateProject(p: Project): Project = {
+        val newProjectList: Seq[NamedExpression] = p.projectList.map {
+          case a: Alias =>
+            a.withNewChildren(Seq(PipeSelect(a.child)))
+              .asInstanceOf[NamedExpression]
```
```suggestion
            a.withNewChildren(Seq(PipeSelect(a.child))).asInstanceOf[NamedExpression]
```

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -0,0 +1,99 @@
+
+-- Expression subqueries in the pipe operator SELECT list.
+table t
+|> select (select a from other where x = a limit 1) as result;
```
can we also test pipe SQL in subquery expression?

dtenedor replied:
Sure, sounds good.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -4989,6 +4989,15 @@ object SQLConf {
+        "the sequence of steps that the query performs in a composable fashion.")
+      .version("4.0.0")
+      .booleanConf
+      .createWithDefault(false)
```
should the default value be `Utils.isTesting`? This feature is opt-in and it won't break any tests.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala:
```diff
@@ -880,4 +880,13 @@ class SparkSqlParserSuite extends AnalysisTest with SharedSparkSession {
+
+  test("Operator pipe SQL syntax") {
+    withSQLConf(SQLConf.OPERATOR_PIPE_SYNTAX_ENABLED.key -> "true") {
+      // Basic selection.
```
hmm, shall we check the parsed plan?

dtenedor replied:
Sounds good, I added some checks. I don't want to compare all the parts of the plan because it's going to be a bit tedious and brittle, but I checked that the essential operators are there :)

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thanks @cloud-fan for your review!!

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala:
```diff
@@ -880,4 +881,20 @@ class SparkSqlParserSuite extends AnalysisTest with SharedSparkSession {
   // scalastyle:on
+
+  test("Operator pipe SQL syntax") {
+    withSQLConf(SQLConf.OPERATOR_PIPE_SYNTAX_ENABLED.key -> "true") {
```
we don't need this now

cloud-fan commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/ThriftServerQueryTestSuite.scala:
```diff
@@ -103,7 +103,8 @@ class ThriftServerQueryTestSuite extends SQLQueryTestSuite with SharedThriftServ
     // SPARK-47264
-    "collations.sql"
+    "collations.sql",
+    "pipe-operators.sql"
```
why it doesn't work in thrift-server?

dtenedor replied:
Update on this, the test flaked out again:

<img width="1122" alt="image" src="https://github.com/user-attachments/assets/ed1f3925-4ca3-4f08-a79f-e065e7750818">

It seems this test suite is not sorting the output result rows in the absence of an ORDER BY clause, like the other SQL test suite variants. I'm adding it back to the blocklist for now.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks all for your reviews :)

dtenedor committed:
re-enable ThriftServerQueryTestSuite blocklist

cloud-fan commented:
thanks, merging to master!
